NKL/Python.lean

```diff
@@ -0,0 +1,288 @@
+deriving instance Repr for Lean.JsonNumber
+
+structure Pos where
+  lineno : Nat
+  end_lineno : Nat := 0
+  col_offset : Nat := 0
+  end_col_offset : Nat := 0
+  deriving Repr
+
+inductive Const where
+  | none
+  | bool (value: Bool)
+  | num (value: Lean.JsonNumber)
+  | string (value: String)
+  | ellipsis
+  deriving Repr
+
+inductive Ctx where
+  | load | store | del
+  deriving Repr
+
+mutual
+inductive Expr where
+  | exprPos (expr : Expr') (pos : Pos)
+  deriving Repr
+
+inductive Expr' where
+  | const (value: Const)
+  | name (id: String) (ctx : Ctx)
+  | attr (value : Expr) (id : String) (ctx : Ctx)
+  | tuple (xs: List Expr) (ctx : Ctx)
+  | list (xs: List Expr) (ctx : Ctx)
+  | subscript (tensor: Expr) (ix: List Expr) (ctx : Ctx)
+  | slice (l u step: Option Expr)
+  | boolOp (op : String) (values : List Expr)
+  | binOp (op : String) (left right : Expr)
+  | unaryOp (op : String) (operand : Expr)
+  | compare (left : Expr) (ops : List String) (comparators : List Expr)
+  | ifExp (test body orelse : Expr)
+  | call (f: Expr) (args: List Expr) (keywords : List Keyword)
+  deriving Repr
+
+inductive Keyword where
+  | keyword (id : String) (value : Expr) (pos : Pos)
+  deriving Repr
+end
+
+mutual
+inductive Stmt where
+  | stmtPos (stmt : Stmt') (pos : Pos)
+  deriving Repr
+
+inductive Stmt' where
+  | pass
+  | expr (e : Expr)
+  | assert (e : Expr)
+  | ret (e: Expr)
+  | assign (xs: List Expr) (e: Expr)
+  | augAssign (x : Expr) (op : String) (e : Expr)
+  | annAssign (x : Expr) (annotation : Expr) (value : Option Expr)
+  | forLoop (x : Expr) (iter: Expr) (body: List Stmt) (orelse : List Stmt)
+  | ifStm (e : Expr) (thn els: List Stmt)
+  deriving Repr
+end
+
+structure Args where
+  posonlyargs : List String
+  args : List String
+  defaults: List Expr
+  vararg : Option String
+  kwonlyargs : List String
+  kw_defaults: List Expr
+  kwarg : Option String
+  deriving Repr
+
+structure Fun where
+  source : String
+  args : Args
+  defaults: List Const
+  body: List Stmt
+  deriving Repr
+
+structure Kernel where
+  entry : String
+  funcs : List (String × Fun)
+  globals : List (String × Option String)
+
+-------------------------------------------------------------------------------
+-- Converting Python AST from Json
+
+namespace Parsing
+open Lean
+
+-- I am using a state monad only to provide better error messages: the source
+-- span (Pos) is saved while traversing the tree to identify the location
+-- of any errors in the original program
+
+abbrev Parser := EStateM String Pos
+
+local instance : MonadLift (Except String) Parser where
+  monadLift
+    | .ok x => return x
+    | .error s => throw s
+
+private def str : Json -> Parser String :=
+  monadLift ∘ Json.getStr?
+
+private def field (f: Json -> Parser a) (j : Json) (name : String) : Parser a :=
+  j.getObjVal? name >>= f
+
+private def field? (f: Json -> Parser a) (j : Json) (name : String) : Parser (Option a) :=
+  try let x <- field f j name; return (some x)
```

Comment:
Is this the same as
```
try field f j name catch _ => return none
```
? If that's right, I'm fine with this, but it seems like a lot of the code is going for maximally point free style.

Reply:
I think the `return (some x)` is needed to get the types right. Maybe `field f j name >>= return . some` :)